osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/mapping/DataObjectToEventListConverter.java

```diff
@@ -64,32 +63,68 @@ private EventDto getEvent(final DataObject eventDataObject, final EventLogCatego
                     "eventData size should be " + eventLogCategory.getNumberOfEventElements());
         }
 
+        if (eventLogCategory == EventLogCategoryDto.POWER_FAILURE_EVENT_LOG) {
+            return getPowerFailureEvent(eventData, eventLogCategory);
+        } else {
+            return getDefaultEvent(eventData, eventLogCategory);
+        }
+    }
+
+    private EventDto getDefaultEvent(List<DataObject> eventData, final EventLogCategoryDto eventLogCategory)
+            throws ProtocolAdapterException {
         // extract values from List<DataObject> eventData.
         final DateTime dateTime = this.extractDateTime(eventData);
         final Short code = this.extractCode(eventData);
         final Integer eventCounter = this.extractEventCounter(eventLogCategory, eventData);
         final String eventLogCategoryName = eventLogCategory.name();
 
-        LOGGER.info("Event time is {}, event code is {}, event category is {} and event counter is {}", dateTime, code,
+        log.info("Event time is {}, event code is {}, event category is {} and event counter is {}", dateTime, code,
                 eventLogCategoryName, eventCounter);
 
         // build a new EventDto with those values.
-        return new EventDto(dateTime, code.intValue(), eventCounter, eventLogCategoryName);
+        return new EventDto(dateTime, code.intValue(), eventCounter, eventLogCategoryName, null, null);
     }
 
-    private DateTime extractDateTime(final List<DataObject> eventData) throws ProtocolAdapterException {
+    private EventDto getPowerFailureEvent(List<DataObject> eventData, final EventLogCategoryDto eventLogCategory)
+            throws ProtocolAdapterException {
+        final DateTime endTime = this.extractDateTime(eventData);
+        final Long duration = this.extractEventDuration(eventData);
+        final String eventLogCategoryName = eventLogCategory.name();
+        final DateTime startTime = calculatePowerFailureStartTime(endTime, duration);
+
+        log.info("PowerFailureEvent startTime is {}, endTime is {}, event duration is {} and event category is {}",
+                startTime,
+                endTime, duration, eventLogCategoryName);
 
+        return new EventDto(endTime, 1, null, eventLogCategoryName, startTime, duration);
+    }
+
+    private DateTime extractDateTime(final List<DataObject> eventData) throws ProtocolAdapterException {
         final DateTime dateTime = this.dlmsHelper.convertDataObjectToDateTime(eventData.get(0)).asDateTime();
+
         if (dateTime == null) {
             throw new ProtocolAdapterException("eventData time is null/unspecified");
         }
+
         return dateTime;
     }
 
-    private Short extractCode(final List<DataObject> eventData) throws ProtocolAdapterException {
+    private Long extractEventDuration(List<DataObject> eventData)
+            throws ProtocolAdapterException {
+        if (!eventData.get(1).isNumber()) {
```

Comment:
possible NPE ?

Reply:
In case of PowerFailureLog there are alway 2 elements. The same is true for al events, minimal of 2.